gapic-generator-java/src/main/java/com/google/api/generator/gapic/composer/common/AbstractTransportServiceStubClassComposer.java

```diff
@@ -1168,4 +1178,254 @@ protected TypeNode getTransportOperationsStubType(Service service) {
 
     return transportOpeationsStubType;
   }
+
+  protected LambdaExpr createRequestParamsExtractorClassInstance(
+      Method method, List<Statement> classStatements) {
+    List<Statement> bodyStatements = new ArrayList<>();
+    VariableExpr requestVarExpr =
+        VariableExpr.withVariable(
+            Variable.builder().setType(method.inputType()).setName("request").build());
+    TypeNode returnType =
+        TypeNode.withReference(
+            ConcreteReference.builder()
+                .setClazz(Map.class)
+                .setGenerics(TypeNode.STRING.reference(), TypeNode.STRING.reference())
+                .build());
+    MethodInvocationExpr.Builder returnExpr =
+        MethodInvocationExpr.builder().setReturnType(returnType);
+    // If the google.api.routing annotation is present(even with empty routing parameters),
+    // the implicit routing headers specified in the google.api.http annotation should not be sent
+    if (method.routingHeaderRule() == null) {
+      createRequestParamsExtractorBodyForHttpBindings(
+          method, requestVarExpr, bodyStatements, returnExpr);
+    } else {
+      createRequestParamsExtractorBodyForRoutingHeaders(
+          method, requestVarExpr, classStatements, bodyStatements, returnExpr);
+    }
+
+    // Overrides extract().
+    // https://github.com/googleapis/gax-java/blob/8d45d186e36ae97b789a6f89d80ae5213a773b65/gax/src/main/java/com/google/api/gax/rpc/RequestParamsExtractor.java#L55
+    return LambdaExpr.builder()
+        .setArguments(requestVarExpr.toBuilder().setIsDecl(true).build())
+        .setBody(bodyStatements)
+        .setReturnExpr(returnExpr.build())
+        .build();
+  }
+
+  private void createRequestParamsExtractorBodyForHttpBindings(
+      Method method,
+      VariableExpr requestVarExpr,
+      List<Statement> bodyStatements,
+      MethodInvocationExpr.Builder returnExprBuilder) {
+    TypeNode routingHeadersBuilderType =
+        TypeNode.withReference(
+            ConcreteReference.builder().setClazz(RequestParamsBuilder.class).build());
+    VariableExpr routingHeadersBuilderVarExpr =
+        VariableExpr.builder()
+            .setVariable(
+                Variable.builder().setName("builder").setType(routingHeadersBuilderType).build())
+            .setIsDecl(true)
+            .build();
```

Comment:
Use RequestParamBuilder to percent-encode the values for http bindings. This use case is for services that need routing headers and backwards-compatibility.